K8s guide

n0tl3ss opened the pull request:
- Missing classes for Kotlin and Groovy
[Kubernetes and the Micronaut Framework.pdf](https://github.com/micronaut-projects/micronaut-guides/files/10034697/Kubernetes.and.the.Micronaut.Framework.pdf)

n0tl3ss committed:
K8s guide

graemerocher commented:
Wow long guide 😄

On first look it looks good, will do a more detailed review soon. We should use Micronaut Serialization for the guide instead of Jackson.

alvarosanchez commented:
@n0tl3ss great guide indeed! 🚀 

After a first scan, I have some thoughts:
- There are parts of the code that might be unnecessary complex for some readers, and perhaps we can simplify them:
  - I would avoid using reactive code.
  - Tests could have inline declarative HTTP clients.
  - We could use Java records for POJOs.
- We could use Dekorate to generate the Kubernetes manifests. There's little value in having users creating them manually, since Dekorate will generate just about the same.
- The `api` Kubernetes service type should be `LoadBalancer`.

n0tl3ss committed:
k8s guide improvements

n0tl3ss committed:
k8s guide improvements

n0tl3ss commented:
Changes:
- Guides add micronaut-serialization feature
- The `api` microservice uses urls from configuration
- Java records instead of POJOs
- Test uses inline declarative HTTP clients
- Only `api` microservice uses reactive code because of 1 CPU pod limitation and we want to move those this to I/O thread it shouldn't be part of main event loop 
- `api` uses LoadBalancer ( It make sense because this is the only service that can scale horizontally).
- minimal java version is 17 because of records usage.

I am against using `Dekorate`, here are reasons:
- it is not supported in Groovy
- Someone who comes to JVM languages from different languages and know k8s well has to learn how to map annotations with `dekorate`.
- With different version of `dekorate` the different Kuberenetes config output might be generated.
- Some thing like adding service account, configuring roles, adding service account to services, etc might not be achievable and has to be done via separate config file or by running commands in Kubernetes CLI in that case there is no much point of using `dekorate` IMO.

PDF:
[Kubernetes and the Micronaut Framework.pdf](https://github.com/micronaut-projects/micronaut-guides/files/10076958/Kubernetes.and.the.Micronaut.Framework.pdf)

n0tl3ss committed:
k8s guide improvements

alvarosanchez commented on buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy:
```diff
@@ -188,6 +189,14 @@ class GuideProjectGenerator implements AutoCloseable {
+                def packageName = BASE_PACKAGE
+                def packageNameWithAppName = packageAndName
+
+                if (app.packageName != null && !app.packageName.isEmpty()) {
```
In Groovy this is simply:

```suggestion
                if (app.packageName) {
```

n0tl3ss committed:
k8s guide improvements

n0tl3ss committed:
add groovy and kotlin soruces

n0tl3ss committed:
kotlin and groovy guides uses classes not the records

n0tl3ss commented:
- Added Kotlin and Groovy sources

Guide maven and groovy:
[Kubernetes and the Micronaut Framework Groovy.pdf](https://github.com/micronaut-projects/micronaut-guides/files/10114679/Kubernetes.and.the.Micronaut.Framework.Groovy.pdf)

Guide gradle and kotlin:
[Kubernetes and the Micronaut Framework Kotlin.pdf](https://github.com/micronaut-projects/micronaut-guides/files/10114678/Kubernetes.and.the.Micronaut.Framework.Kotlin.pdf)

n0tl3ss committed:
Add error example

alvarosanchez commented on guides/micronaut-k8s/api/groovy/src/main/groovy/example/micronaut/ErrorExceptionHandler.groovy:
```diff
@@ -0,0 +1,17 @@
+import io.micronaut.http.server.exceptions.ExceptionHandler
+import jakarta.inject.Singleton
+
+@Produces
```
Do we need this empty `@Produces`?

alvarosanchez commented on guides/micronaut-k8s/api/groovy/src/main/groovy/example/micronaut/auth/Credentials.groovy:
```diff
@@ -0,0 +1,9 @@
+
+import io.micronaut.context.annotation.ConfigurationProperties
+
+@ConfigurationProperties("micronaut.authentication-credentials")
```
IMHO is better to educate users to use a different namespace for configuration properties. `micronaut.*` should only be used by Micronaut itself.

alvarosanchez commented on guides/micronaut-k8s/api/groovy/src/main/groovy/example/micronaut/controllers/GatewayController.groovy:
```diff
@@ -0,0 +1,84 @@
+
+    @Get("/users/{id}") // <2>
+    Mono<User> getUserById(@NonNull Integer id) {
+         userClient.getById(id).publishOn(Schedulers.boundedElastic())
```
I still think using reactive overcomplicates the code, where you could simply change everything to be blocking and use `@ExecuteOn(TaskExecutors.IO)`

alvarosanchez commented on guides/micronaut-k8s/api/groovy/src/main/groovy/example/micronaut/models/Order.groovy:
```diff
@@ -0,0 +1,45 @@
+
+    @NotBlank @Nullable @JsonProperty("user_id") Integer userId
+
+    @JsonProperty @Nullable User user
```
Do we need these default `@JsonProperty`s?

alvarosanchez commented on guides/micronaut-k8s/api/java/src/main/java/example/micronaut/auth/Credentials.java:
```diff
@@ -0,0 +1,10 @@
+        String username,
+        String password
+        ) {
+        }
```
```suggestion
}
```

alvarosanchez commented on guides/micronaut-k8s/api/java/src/main/java/example/micronaut/models/Item.java:
```diff
@@ -0,0 +1,17 @@
+
+@Serdeable
+public record Item(
+        @JsonProperty("id") Integer id,
```
Do we need those `@JsonProperty`s?

alvarosanchez commented on guides/micronaut-k8s/api/src/main/resources/application-dev.yml:
```diff
@@ -0,0 +1,8 @@
+netty:
+  default:
+    allocator:
+      max-order: 3
```
Can be removed?

alvarosanchez commented on guides/micronaut-k8s/api/src/main/resources/bootstrap-dev.yml:
```diff
@@ -0,0 +1,17 @@
+  application:
+    name: api
+  config-client:
+    enabled: false # <1>
```
false is the default. Only add it to set it to true

alvarosanchez commented on guides/micronaut-k8s/api/src/main/resources/bootstrap-dev.yml:
```diff
@@ -0,0 +1,17 @@
+kubernetes:
+  client:
+    secrets:
+      enabled: false # <4>
```
false is the default. Only add it to set it to true

alvarosanchez commented on guides/micronaut-k8s/api/src/test/resources/application-test.yml:
```diff
@@ -0,0 +1,8 @@
+netty:
+  default:
+    allocator:
+      max-order: 3
```
Remove?

alvarosanchez commented on guides/micronaut-k8s/api/src/test/resources/bootstrap-test.yml:
```diff
@@ -0,0 +1,9 @@
+micronaut:
```
Whole file can be removed

alvarosanchez commented on guides/micronaut-k8s/metadata.json:
```diff
@@ -0,0 +1,27 @@
+    {
+      "name": "users",
+      "features": ["discovery-kubernetes", "management", "security", "kubernetes", "serialization-jackson"],
+      "packageName": "example.micronaut"
```
This is the default, isn't it?

n0tl3ss replied:
I added code that if you specify `packageName` inside application block it will use  `name` field as application name. By default all apps are named `micronautguide` and in this example it is required to be named differently because of docker images. I did it in this way to keep it backward compatible with older guides and now other guides if someone wants can name package differently from `example.micronaut`.

n0tl3ss committed:
Update buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy

Co-authored-by: Álvaro Sánchez-Mariscal <alvaro.sanchez-mariscal@oracle.com>

n0tl3ss committed:
guide improvements

n0tl3ss committed:
guide improvements

n0tl3ss committed:
guide improvements

n0tl3ss committed:
guide improvements

n0tl3ss commented:
Maven Java guide:

[Kubernetes and the Micronaut Framework (Java Maven).pdf](https://github.com/micronaut-projects/micronaut-guides/files/10123725/Kubernetes.and.the.Micronaut.Framework.Java.Maven.pdf)

The guides needs a new version of [micronaut-kuberentes](https://github.com/micronaut-projects/micronaut-kubernetes) if you want to test it, please use `3.4.1-SNAPSHOT` version

n0tl3ss committed:
update tags

alvarosanchez commented on guides/micronaut-k8s/orders/java/src/main/java/example/micronaut/auth/CredentialsChecker.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public Publisher<AuthenticationResponse> authenticate(@Nullable HttpRequest<?> httpRequest,
+                                                          AuthenticationRequest<?, ?> authenticationRequest) {
+        return Flux.create(emitter -> {
```
This should be a `Mono`. See https://micronaut-projects.github.io/micronaut-security/snapshot/guide/#authenticationProviders for an example.

alvarosanchez commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,880 @@
+
+test:UsersClient[app=users]
+
+`HealthTest.kt` checks if there is `/health` endpoint exposed that is needed for service discovery.
```
This depends on the language

alvarosanchez commented on guides/micronaut-k8s/users/java/src/main/java/example/micronaut/models/User.java:
```diff
@@ -0,0 +1,17 @@
+import javax.validation.constraints.Max;
+import javax.validation.constraints.NotBlank;
+
+@Serdeable
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/users/java/src/main/java/example/micronaut/auth/Credentials.java:
```diff
@@ -0,0 +1,10 @@
+
+import io.micronaut.context.annotation.ConfigurationProperties;
+
+@ConfigurationProperties("authentication-credentials")
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/users/java/src/main/java/example/micronaut/auth/CredentialsChecker.java:
```diff
@@ -0,0 +1,35 @@
+import reactor.core.publisher.FluxSink;
+
+@Singleton
+public class CredentialsChecker implements AuthenticationProvider {
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/users/java/src/test/java/example/micronaut/UsersClient.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.List;
+
+@Client("/")
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/users/java/src/test/java/example/micronaut/HealthTest.java:
```diff
@@ -0,0 +1,26 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@MicronautTest
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/users/src/main/resources/bootstrap-dev.yml:
```diff
@@ -0,0 +1,7 @@
+micronaut:
+  application:
+    name: users
```
Can be removed

alvarosanchez commented on guides/micronaut-k8s/api/java/src/main/java/example/micronaut/controllers/GatewayController.java:
```diff
@@ -0,0 +1,84 @@
+
+@Controller("/api") // <1>
+@Validated
+@ExecuteOn(TaskExecutors.IO)
```
Add existing callout

alvarosanchez commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,880 @@
+[source,text]
+----
+NAME         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
+api          NodePort    10.102.170.41    <none>        8080:32757/TCP   28m
```
This should be a LoadBalancer, and you should have an external IP

alvarosanchez commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,880 @@
+[source,bash]
+----
+export API_URL=$(minikube service api --url)
+----
```
Not sure in Minikube, but in other distros, creating the service as `LoadBalancer` will expose the service in localhost. Please check if this is still needed.

n0tl3ss replied:
I added note how you can expose ip if someone wants, by default it will be in pending state.

n0tl3ss committed:
guide improvements

n0tl3ss commented:
New version:
[Kubernetes and the Micronaut Framework (Maven).pdf](https://github.com/micronaut-projects/micronaut-guides/files/10133688/Kubernetes.and.the.Micronaut.Framework.Maven.pdf)

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+
+Let's describe the microservices you will build through the guide.
+
+* `users` - This microservice contains all customers data that can place orders on items, also a new customer can be created. Microservice requires Basic authentication to access it.
```
we can rename `users` microservice into `customers`. Authentication is required to showcase Kubernetes distributed configuration feature between microservices.

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+:exclude-for-build:gradle
+
+[source, bash]
+.orders
```
```suggestion
.users
```

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Apply suggestions from code review

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+:exclude-for-build:maven
+
+[source, bash]
+.users
```
```suggestion
.api
```

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+:exclude-for-build:gradle
+
+[source, bash]
+.users
```
```suggestion
.api
```

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+:exclude-for-build:maven
+
+[source, bash]
+.orders
```
```suggestion
.api
```

n0tl3ss commented on guides/micronaut-k8s/micronaut-k8s.adoc:
```diff
@@ -0,0 +1,898 @@
+:exclude-for-build:gradle
+
+[source, bash]
+.orders
```
```suggestion
.api
```

n0tl3ss committed:
Apply suggestions from code review

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss committed:
Update guides/micronaut-k8s/micronaut-k8s.adoc

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

n0tl3ss commented on guides/micronaut-k8s/orders/java/src/main/java/example/micronaut/models/Order.java:
```diff
@@ -0,0 +1,20 @@
+package example.micronaut.models;
```
That is expected. In `api` microservice  Order class is extended to contain information about user that placed the order.

n0tl3ss committed:
code review improvements

n0tl3ss committed:
Merge branch 'feature/k8s-guide' of https://github.com/micronaut-projects/micronaut-guides into feature/k8s-guide

timyates commented:
Working through this, the users pod doesn't start up as:

```
16:38:53.800 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: io/micronaut/jackson/env/JsonPropertySourceLoader        │
│ java.lang.NoClassDefFoundError: io/micronaut/jackson/env/JsonPropertySourceLoader                                                                    │
│     at io.micronaut.kubernetes.util.KubernetesUtils.<clinit>(KubernetesUtils.java:61)                                                                │
│     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySourcesFromConfigMaps(KubernetesConfigurationClient.java:169)  │
│     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySources(KubernetesConfigurationClient.java:110)                │
│     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.lambda$getPropertySources$0(DefaultCompositeConfigurationClient.java:64)    │
│     at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)                                                           │
│     at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)                                                     │
│     at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)                                                               │
│     at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)                                                        │
│     at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)                                                          │
│     at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)                                                               │
│     at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)                                                              │
│     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.getPropertySources(DefaultCompositeConfigurationClient.java:65)             │
│     at io.micronaut.discovery.client.config.DistributedPropertySourceLocator.findPropertySources(DistributedPropertySourceLocator.java:75)           │
│     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.readPropertySourceList(DefaultApplicationContext.java:783)        │
│     at io.micronaut.context.env.DefaultEnvironment.readPropertySources(DefaultEnvironment.java:412)                                                  │
│     at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:270)                                                                │
│     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:769)                         │
│     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:738)                         │
│     at io.micronaut.context.DefaultApplicationContext.startEnvironment(DefaultApplicationContext.java:242)                                           │
│     at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:193)                                                      │
│     at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)                                                                                       │
│     at io.micronaut.runtime.Micronaut.run(Micronaut.java:323)                                                                                        │
│     at io.micronaut.runtime.Micronaut.run(Micronaut.java:309)                                                                                        │
│     at example.micronaut.Application.main(Application.java:7)                                                                                        │
│ Caused by: java.lang.ClassNotFoundException: io.micronaut.jackson.env.JsonPropertySourceLoader                                                       │
│     at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)                                                       │
│     at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)                                                    │
│     at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
```

n0tl3ss commented:
> Working through this, the users pod doesn't start up as:
> 
> ```
> 16:38:53.800 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: io/micronaut/jackson/env/JsonPropertySourceLoader        │
> │ java.lang.NoClassDefFoundError: io/micronaut/jackson/env/JsonPropertySourceLoader                                                                    │
> │     at io.micronaut.kubernetes.util.KubernetesUtils.<clinit>(KubernetesUtils.java:61)                                                                │
> │     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySourcesFromConfigMaps(KubernetesConfigurationClient.java:169)  │
> │     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySources(KubernetesConfigurationClient.java:110)                │
> │     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.lambda$getPropertySources$0(DefaultCompositeConfigurationClient.java:64)    │
> │     at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)                                                           │
> │     at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)                                                     │
> │     at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)                                                               │
> │     at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)                                                        │
> │     at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)                                                          │
> │     at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)                                                               │
> │     at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)                                                              │
> │     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.getPropertySources(DefaultCompositeConfigurationClient.java:65)             │
> │     at io.micronaut.discovery.client.config.DistributedPropertySourceLocator.findPropertySources(DistributedPropertySourceLocator.java:75)           │
> │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.readPropertySourceList(DefaultApplicationContext.java:783)        │
> │     at io.micronaut.context.env.DefaultEnvironment.readPropertySources(DefaultEnvironment.java:412)                                                  │
> │     at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:270)                                                                │
> │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:769)                         │
> │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:738)                         │
> │     at io.micronaut.context.DefaultApplicationContext.startEnvironment(DefaultApplicationContext.java:242)                                           │
> │     at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:193)                                                      │
> │     at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)                                                                                       │
> │     at io.micronaut.runtime.Micronaut.run(Micronaut.java:323)                                                                                        │
> │     at io.micronaut.runtime.Micronaut.run(Micronaut.java:309)                                                                                        │
> │     at example.micronaut.Application.main(Application.java:7)                                                                                        │
> │ Caused by: java.lang.ClassNotFoundException: io.micronaut.jackson.env.JsonPropertySourceLoader                                                       │
> │     at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)                                                       │
> │     at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)                                                    │
> │     at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
> ```

You have to use latest  'SNAPSHOT' version of micronaut k8s module. With 3.8.2 this should work out of the box.

timyates commented:
> > Working through this, the users pod doesn't start up as:
> > ```
> > 16:38:53.800 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: io/micronaut/jackson/env/JsonPropertySourceLoader        │
> > │ java.lang.NoClassDefFoundError: io/micronaut/jackson/env/JsonPropertySourceLoader                                                                    │
> > │     at io.micronaut.kubernetes.util.KubernetesUtils.<clinit>(KubernetesUtils.java:61)                                                                │
> > │     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySourcesFromConfigMaps(KubernetesConfigurationClient.java:169)  │
> > │     at io.micronaut.kubernetes.configuration.KubernetesConfigurationClient.getPropertySources(KubernetesConfigurationClient.java:110)                │
> > │     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.lambda$getPropertySources$0(DefaultCompositeConfigurationClient.java:64)    │
> > │     at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)                                                           │
> > │     at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)                                                     │
> > │     at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)                                                               │
> > │     at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)                                                        │
> > │     at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)                                                          │
> > │     at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)                                                               │
> > │     at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)                                                              │
> > │     at io.micronaut.discovery.config.DefaultCompositeConfigurationClient.getPropertySources(DefaultCompositeConfigurationClient.java:65)             │
> > │     at io.micronaut.discovery.client.config.DistributedPropertySourceLocator.findPropertySources(DistributedPropertySourceLocator.java:75)           │
> > │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.readPropertySourceList(DefaultApplicationContext.java:783)        │
> > │     at io.micronaut.context.env.DefaultEnvironment.readPropertySources(DefaultEnvironment.java:412)                                                  │
> > │     at io.micronaut.context.env.DefaultEnvironment.start(DefaultEnvironment.java:270)                                                                │
> > │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:769)                         │
> > │     at io.micronaut.context.DefaultApplicationContext$RuntimeConfiguredEnvironment.start(DefaultApplicationContext.java:738)                         │
> > │     at io.micronaut.context.DefaultApplicationContext.startEnvironment(DefaultApplicationContext.java:242)                                           │
> > │     at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:193)                                                      │
> > │     at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)                                                                                       │
> > │     at io.micronaut.runtime.Micronaut.run(Micronaut.java:323)                                                                                        │
> > │     at io.micronaut.runtime.Micronaut.run(Micronaut.java:309)                                                                                        │
> > │     at example.micronaut.Application.main(Application.java:7)                                                                                        │
> > │ Caused by: java.lang.ClassNotFoundException: io.micronaut.jackson.env.JsonPropertySourceLoader                                                       │
> > │     at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)                                                       │
> > │     at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)                                                    │
> > │     at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
> > ```
> 
> You have to use latest 'SNAPSHOT' version of micronaut k8s module. With 3.8.2 this should work out of the box.

Awesome, thanks 👍

sdelamo commented:
close via https://github.com/micronaut-projects/micronaut-guides/pull/1175

n0tl3ss committed:
Draft guide for OCI Kubernetes (#1175)

* Switch to use Serde for the S3 Event guide (#1138)

* Switch to use Serde for the Discovery eureka guide (#1134)

* Switch to use Serde for the Oauth client credentials auth0 guide (#1133)

* Switch to use Serde for the Discovery consul guide (#1135)

* Use the correct version for the micronaut-data processor with SpringBoot (#1164)

We were using the version of Micronaut, which is incorrect

* Guide: Connect a Micronaut JMS Application to an AWS SQS Queue (#1160)

Max supported Java version by AWS SDK 1.x is 16 (https://github.com/aws/aws-sdk-java#maintenance-and-support-for-java-versions)

* add AWS SQS twitter card

* Remove references to postgres and use mysql consistently for guide (#1166)

closes #1161

* Switch to use Serde for the Mybatis data access guide (#1139)

* Switch to use Serde for the Mybatis data access guide

* Remove Creator, fix Kotlin warning

* Fixes missing section in trace guide (#1173)

* Fixes wrong callouts to WarehouseController endpoints (#1174)

* Draft guide for OCI Kubernetes

* Apply suggestions from code review

Co-authored-by: Bernard Horan <bernard.horan@oracle.com>

* Update guides to 3.7.5 (#1179)

* code review improvements add native image support

* code review improvements add native image support

* code review improvements

* code review improvements

* Build Guides with Micronaut 3.8.0 (#1157)

* Build Guides with Micronaut 3.8.0

This PR runs the guides with Micronaut 3.8.0-SNAPSHOT as a pre-release check

Removes the local localstack feature as this now exists in 3.8.0

* Add snapshot repo for snapshot versions of Micronaut

* Look up the version for the data processor

* version 3.8.0

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

* Remove ambiguity for Kotlin (#1182)

It couldn't work out which mathod to call as the closure had an implicit return value.

I'm not sure why this has only recently appeared...  Maybe a change to micrometer

* Fix inputs for guides with a base (#1183)

If we have a guide that declares a base project, ie:

https://github.com/micronaut-projects/micronaut-guides/blob/dd911994e0c54c068602389dbcbe08eac8bdf045/guides/micronaut-metrics-aws/metadata.json#L2

Then we were not taking the code from micronaut-metrics into account for the cache-key, so changes were not picked up when building micronaut-metrics-aws

This change adds the base directory as an input if it was set

* remove empty constructor (#1167)

* misc fixes and cleanup (#1186)

* Add R2DBC guide (#965)

* copy: change GraalVM version (#1187)

* change publication to 2023-01-04

* fix: category should be categories

* disable spring-boot-micronaut-data guide (#1185)

* Micronaut Framework 3.8.1 (#1195)

* Kotlin and Groovy versions of Micronaut Dependency Injection types (#1227)

Close #1082

* Add K8s AWS guide

* Add K8s AWS guide

* Kotlin version of JAX-RS guide (#1225)

Close #1221

* code review improvements

* Micronaut Framework 3.8.2 (#1229)

* fix java version typo (#1234)

* Mock Github Server to avoid rate limiting in HTTP Client Guide (#1236)

* build: Micronaut Framework 3.8.4

* X-API-KEY guide (#1237)

* add api key guide twitter card

* CORS guide (#1238)

* docker image guide (#1239)

* executable Jar guide to distribution category (#1240)

- enable every language. Use distributable-base

* create the http client lazyly (#1243)

* add lateinit and @field callouts (#1228)

* Custom constraint guide (#1244)

* custom constraint

* Update guides/micronaut-custom-validation-annotation/java/src/test/java/example/micronaut/ContactTest.java

* twitter card

* Fix curl command (#1248)

When copying the curl command, each new line is executed rather than waiting for the last command.

* build: Micronaut Framework 3.8.5(#1249)

* Guide: Using IntelliJ IDEA to develop Micronaut applications (#1241)

* 2023 copyright

* Micronaut Framework 3.8.6 (#1251)

* build: Micronaut Framework 3.8.7 (#1252)

* add awaitility version (#1253)

Close: https://github.com/micronaut-projects/micronaut-guides/issues/1074

* Fixes for MN Scope Types guide, mostly imports and annotations to ensure that the Tests work as intended. (#1254)

* [Error-handling Guide] Added missing dependency and a closing bracket for the BookController class. (#1260)

* Micronaut Framework 3.8.8 (#1261)

* move create your first graalvm guide to distribution category (#1256)

* Security Session Guide: Removed the extra dependency that could cause tests to fail (#1263)

* CRaC Cloud Run guide (#1259)

Close #1257

* Keycloak Guide (#586)

* add back common-google-cloud-platform.adoc snippet

accidentally deleted via https://github.com/micronaut-projects/micronaut-guides/pull/1259

* add image back

* use categories

* Micronaut Framework 3.8.9 (#1264)

* [Database JDBC Guides for AWS, Azure, and GCP] Fixed the environment variable names to be compatible with the sample JDBC application, and added explanations similar to the original JDBC guide. (#1270)

* build: Micronaut Framework 3.9.1 (#1267)

* build: Micronaut Build Plugin to 5.4.9 (#1273)

* 3.9.2 (#1280)

---------

Co-authored-by: Tim Yates <tim.yates@gmail.com>
Co-authored-by: Slavko <bodvanski.slavko@gmail.com>
Co-authored-by: sdelamo <sergio.delamo@softamo.com>
Co-authored-by: Dean Wette <wetted@objectcomputing.com>
Co-authored-by: Bernard Horan <bernard.horan@oracle.com>
Co-authored-by: Burt Beckwith <burt@burtbeckwith.com>
Co-authored-by: Graeme Rocher <graeme.rocher@gmail.com>
Co-authored-by: Alina Yurenko <alina.yurenko@oracle.com>
Co-authored-by: hartraft <hartraft@tutanota.com>
Co-authored-by: Long Pham <ptlitmipitu@gmail.com>

sdelamo committed:
Merge branch 'master' into feature/k8s-guide
